# 기본 자료형: 함수

_[Think Python의 3장](http://greenteapress.com/thinkpython2/html/thinkpython2004.html) 
내용을 요약 및 수정한 내용입니다._

## 수학에서의 함수

함수라는 표현이 수학에서 많이 사용된다. 
수학에서 함수는 두 집합 사이의 관계이며,
첫째 집합의 원소를 둘째 집합의 원소와 대응시킨다. 

아래 그림은 집합 $X$의 원수 $x$와 집합 $Y$의 원소 $f(x)$를 대응시키는 함수를 보여준다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/fun_math1.png">
</td>
</tr>
</table>
</p>

<그림 출처: [함수, 위키백과](https://ko.wikipedia.org/wiki/함수)>

이때 $X$와 $Y$를 각각 함수 $f$의 **정의역**(domain)과 
**공역**(codomain)이라 부른다.

## 프로그래밍에서의 함수

프로그래밍에서 함수가 하는 역할은 다음과 같다. 

> 어떤 값이 입력(input)되면 그 값을 이용하여 특정 수식에 따라 계산한 후 
    특정 값을 내준다(output).
  
아래 그림은 함수의 입력(input)과 내주기(output)의 관계를 보여준다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/fun_prog2.png">
</td>
</tr>
</table>
</p>

<그림 출처: [함수, 위키백과](https://ko.wikipedia.org/wiki/함수)>

함수의 입력값으로 사용될 수 있는 값들의 집합이 정의역에 해당하고,
내주는 값들의 집합이 공역에 해당한다.
함수가 내주는 값을 **리턴값**이라 부르기도 한다.
이유는 함수가 내주는 값을 보통 `return` 지정자로 지정하기 때문이다.

### 차이점

입력값을 특정 집합의 원소로 보고, 리턴값을 다른 집합의 원소로 보고,
그리고 
두 집합 사이의 대응관계를 "수식에 따라 계산한다"라 이해한다면 
수학과 프로그래밍에서의 함수 개념은 동일하다.

* 정의역: 함수의 입력값들로 구성된 집합
* 공역: 함수의 리턴값들로 구성된 집합

실제로, 수학 함수를 프로그래밍 함수로 다룰 수 있다. 
예를 들어, 절댓값 함수는 실수에서 실수로 가는 함수이며,
양수는 그대로 두고, 음수는 부호를 바꿔 양수로 만든다. 

절댓값에 해당하는 프로그래밍 함수는 `abs`이며, 아래와 같이 작동한다.

In [12]:
abs(-3.3)

3.3

위에서 사용된 표현식 `abs(-3.3)`과 그 결과인 `3.3`의 역할은 다음과 같다.

* `abs`: 절댓값 함수를 가리키는 이름
* `-3.3`: 입력값
* `3.3`: 리턴값

#### 프로그래밍 함수의 정의역

앞서 프로그래밍 함수의 입력값들로 구성된 집합이 수학 함수의 정의역에 
해당한다고 설명하였다.
그런데 주의사항이 하나 있다.

파이썬의 경우, 함수의 입력값에 어떤 제한도 없기 때문이다.
다만, 의도하지 않은 값이 사용될 경우, 함수의 실행 과정에서 오류가 발생한다.
예를 들어, `abs` 함수를 문자열과 함께 실행하면 
`TypeError`, 즉, 인자의 자료형에 문제가 있다는 오류가 발생한다.
이유는 `abs` 함수가 숫자만 다룰 수 있기 때문이다.

In [13]:
abs('python')

TypeError: bad operand type for abs(): 'str'

따라서 프로그래밍 함수의 정의역은 입력값들 중에서 오류가 발생하지 않는 값들의 집합으로
보는 게 보다 더 적절하다.
하지만, 이것을 프로그래밍 함수의 정의역으로 **정의**(define)하지는 않을 것이다.
프로그래밍 함수의 정의역을 엄밀하게 정의하는 일은 보다 복잡하기에,
자세히 다루지 않는다. 
다만, 개념적으로 프로그래밍 함수와 수학 함수가 유사한 면이 많다라는 정도만
기억해두면 좋다.

### 프로그래밍 함수 예제

프로그래밍에서는 다루는 함수 유형을 살펴본다.

#### 수학 함수와 유사한 경우

* 수를 입력하면 계산결과를 돌려준다.

In [14]:
def toto(x):
    y = x**2+1
    return y
    
print(toto(3))

10


* 아래 함수는 변수이름 `y`를 지역변수와 전역변수로 동시에 사용했다.
    하지만 서로 관계가 없기 때문에 일반 수학 함수처럼 작동한다.

In [15]:
y = -1

def toto(x):
    y = x**2+1
    return y

# 전역변수 y는 함수와 아무 상관 없음.
print(y, toto(3))

-1 10


In [16]:
y = -1

def toto(x):
    y = x**2+1
    return y

# 전역변수 y는 함수와 아무 상관 없음.
print(toto(3), y)

10 -1


#### 수학 함수와 조금 다른 경우

* 아래 함수는 함수 안에서 전역변수 `y`를 업데이트한다.
    이렇게 하면 `y`가 가리키는 값은 `toto` 함수의 호출에
    민감하게 반응한다.

In [9]:
y = -1

def toto(x):
    global y
    y = x**2+1
    return y

# toto 함수가 실행되면 y 가 업데이트 됨.
print(toto(3), y)

10 10


In [10]:
y = -1

def toto(x):
    global y
    y = x**2+1
    return y

# y 값을 먼저 확인 한 후에 toto 함수가 실행
print(y, toto(3))

-1 10


#### 수학 함수와 많이 다른 경우

* 함수의 리턴값이 지정되지 않으면 `None`을 내줌.
* `None`은 아무런 의미가 없는 값임.
* 대신에 아래 함수를 실행하면 변수 `y`의 값을 업데이트 함.
* 수학엔 이런 함수 없음.

In [34]:
y = -1

def toto(x):
    global y
    y = x**2+1
    
print(toto(3), y)

None 10


In [11]:
y = -1

def toto(x):
    global y
    y = x**2+1
    
print(y, toto(3))

-1 None


## 값과 함수

프로그램을 구현한다는 것은 기본적으로 특정 종류의 값들을 조작하여 원하는 값을 구하는 과정을
특정 프로그래밍 언어로 작성하는 것이다. 

지금가지 살펴 본 값들의 종류는 다음과 같다.
* 정수: -1, 0, 2, -3, ...
* 실수: 0.25, -3.54, ...
* 불리언 값: True, False
* 문자열: "python", "a3ge", ...
* 함수 등등

모든 프로그래밍 언어마다 값을 표현하는 고유의 방식이 있다.
정수와 실수는 거의 모든 언어에서 표현하는 방식이 동일하지만 이외의 자료형에 속하는 값들은 비슷하지만 조금씩 다르게
구현되었으며, 따라서 사용되는 방식도 그에 따라 조금씩 다를 수 있다.

따라서 특정 자료형을 어떻게 표현하는지 정확하게 알고 있어야 하며,
문법에 맞지 않게 사용할 경우 기본적으로 오류가 발생한다. 
여기서는 함수를 이용하여 값을 표현하는 방식을 알아본다. 

먼저 함수가 무엇인지를 알아본다.

### 함수와 매개변수

함수의 정의는 아래 형식을 갖는다.

```python
def 함수이름(매개변수1, 매개변수2, ..., 매개변수n):
    명령문
    return 표현식
```

**주의:** 함수 본체에 사용되는 명령문은 들여쓴다.

함수를 정의할 때 사용되는 **매개변수**는 인자로 들어오는 값들을
함수 본체 명령문에 전달하는 역할을 수행한다.

#### 함수 정의 예제

절댓값을 계산하는 함수 `myAbs`와
두 숫자의 합을 계산하는 함수 `myAdd`를 아래와 같이 직접 정의할 수 있다.

`myAbs`는 한 개의 값을 입력받아야 하기에 한 개의 매개변수가 필요하고,
반면에 `myAdd`는 더해야 할 두 개의 값을 입력받아야 하기에 두 개의 매개변수가 필요하다.

In [20]:
def myAbs(num):
    if num >= 0:
        z = num
    else:
        z = -num
    return z

def myAdd(num1, num2):
    z = num1 + num2
    return z

In [21]:
myAbs(-3)

3

In [22]:
myAdd(-3, myAbs(-3))

0

### 매개변수와 인자

`myAbs`와 `myAdd`의 정의헤서 사용된 
`num`, `num1`, `num2` 등은 함수의 인자를 받아들이는데 사용되는 **매개변수**이다.
그리고 매개변수를 통해 함수에게 전달되는 값들을 **인자**라고 부른다. 
사용되는 인자의 개수는 매개변수의 개수와 일치해야 한다.

함수와 매개변수들의 이름은 각각의 역할에 맞게 정하는 것을 권유한다.
그러면 함수와 매개변수들의 이름을 보고 함수와 각 매개변수들의 의미와 
역할을 파악하는 데에 보다 유리하다.

### 옵션 인자

`print` 함수를 이용하여 화면에 출력할 때 사용되는 인자의 수는 여러 개일 수 있다.

In [23]:
print('Hello,', 'Python', '!')

Hello, Python !


그런데 여러 개의 인자를 각각 다른 줄에 출력하려면 아래와 같이 하면 된다.

In [25]:
print('Hello,', 'Python', '!', sep='\n')

Hello,
Python
!


위에서 사용된 `sep`은 `print` 함수의 옵션 인자이다.
`sep`과 같은 옵션 인자는 인자를 지정하지 않으면 기본값을 사용한다.
`sep`의 기본값은 한 칸 띄어쓰기를 의미하는 `' '`이다.
즉, `print` 함수의 인자들을 한 칸씩 띄어서 화면에 보여준다.
그리고 위에서는 `sep`에 대한 인자를 띄어쓰기 대신 줄바꿈(`'\n'`)을
사용하였다.

이렇듯 특정 함수들은 옵션 인자를 사용할 수 있으며, 그런 함수는
매개변수에 기본값을 지정하는 식으로 정의된다.

예를 들어, `print`함수에 사용되는 매개변수 중에
`sep` 이외에 `end`, `file`, `flush` 등이 기본값을 갖는다.
실제로 `help(print)` 명령문을 실행하면
아래와 같이 확인된다.

```python
print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
```

* `sep` 매개변수: 출력할 인자들 사이에 대한 기준 지정. 기본값은 띄어쓰기.
* `end` 매개변수: 화면 출력 후 추가로 할 일에 대한 기준 지정. 기본값은 줄바꿈.
* `file` 매개변수: 출력 장치 지정. 기본값은 터미널 화면.
* `flush` 매개변수: 여러 개의 출력값들을 하나씩 차례대로 출력 장치에 보낼지 말지를 지정. 기본값은 하나씩 바로바로 보내기.

위 옵션 매개변수 중에서 `sep`과 `end`는 여러 모로 유용하게 활용할 수 있다.

In [29]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



### 예제

옵션 인자를 사용하는 함수를 정의할 수 있다.
아래 함수는 두 개의 숫자를 입력받아 합을 구한다.

In [32]:
def myAdd10(left, right=10):
    add10 = left + right
    return add10

그런데, 둘째 매개변수의 기본값이 10으로 지정되었다.
따라서 둘째 인자를 반드시 입력하지 않아도 되며,
그럴 경우 둘째 인자는 10으로 처리된다.

In [33]:
myAdd10(5)

15

물론 둘째 인자를 지정할 수도 있다.
옵션 매개변수의 값을 임의로 지정하고 싶으면 
옵션 매개변수의 이름을 언급하는 것이 좋다.

In [35]:
myAdd10(5, 20)

25

In [36]:
myAdd10(5, right=20)

25

## 동적 타이핑 vs. 정적 타이핑

함수는 임의의 값을 인자로 받는다. 
하지만 함수마다 다룰 수 있는 값들의 자료형이 존재하며,
다룰 수 없는 자료형의 값이 인자로 사용되면 오류가 발생한다.

예를 들어, `type` 함수는 모든 자료형의 값을 다룰 수 있지만,
`abs` 함수는 정수와 실수는 다루지만, 문자열을 
인자로 사용하면 오류가 발생한다. 

In [37]:
abs('-3.3')

TypeError: bad operand type for abs(): 'str'

이렇듯, 파이썬 함수는 임의의 값에 적용될 수 있지만,
인자의 자료형을 파이썬 해석기가 파악해서 적용가능 여부를 판단한다.
이런 기능을 **동적 타이핑**이라 부른다.
즉, 처음부터 인자의 자료형을 지정하거나 확인하는 게 아니라
프로그램을 실행하는 도중에 사용되는 함수 인자들의 자료형을 확인한다는 의미이다.
이는 함수가 다루는 모든 값에 대해 동일하다. 

### 파이썬과 정적 타이핑

C, Java 등 많은 프로그래밍 언어는 동적 타이핑 대신에 **정적 타이핑**(static typing)을 지원한다.
즉, 함수나 변수를 선언할 때 사용되는 변수들의 자료형과 인자 및 반환값의 자료형을 
애초부터 명시해야 하며 지정된 자료형이 사용되지 않을 경우 오류를 발생시킨다.

파이썬은 3.6 버전부터 정적 타이핑 형식을 지원한다. 
다만 C, Java의 자료형과 관련된 엄격함은 전혀 존재하지 않으며, 
그냥 정적 타이핑의 형식만 빌려왔다.

즉, **자료형 명시**(type annotations)를 지원할 뿐이며, 
실제로는 동적 타이핑 형식으로 문법과 실행 과정을 확인하고 제어한다.

예를 들어, `add` 함수를 아래와 같이 선언할 수 있다.

In [38]:
def myAdd(a: float, b: float) -> float:
    return a + b

`myAdd` 함수는 부동소수점의 합을 계산함수라는 의도를 
명확히 하기 위해 `float`라는 자료형을 입력값와 리턴값에 대해 명시하였다.

그래서 위 함수는 마치 부동소수점에 대해서만 작동하는 것처럼 보인다.
하지만 이는 `myAdd` 함수는 여전히 다른 자료형의 값들을 
인자로 사용할 수 있다.

In [40]:
print(myAdd(10, 5))
print(myAdd([1, 2], [3]))
print(myAdd("저 ", "잠깐만요!"))

15
[1, 2, 3]
저 잠깐만요!


즉, 자료형의 명시는 함수 정의의 의도를 전달하는 역할만 수행하며,
파이썬이 정적타이피을 지원한다는 의미는 아님에 주의해야 한다.

### 지금까지 사용한 함수

앞서 살펴 본 수학함수 이외에 많은 함수들이 있으며 
이미 여러 함수들을 살펴보았다.
`type`, `int`, `float`, `str`, `input`, `time.sleep`, `urllib.request.urlopen`, 등등

In [4]:
type(42)

int

In [5]:
int('32')

32

In [6]:
float('3.9999')

3.9999

In [7]:
str(32)

'32'

In [8]:
type(str(32))

str

### 함수와 추상화
도구가 이미 준비되어 있을 때 그 도구가 어떻게 만들어졌는지 알 필요 없듯이
이미 정의된 함수가 어떻게 정의되어 있는가 또한 굳이 알 필요 없다.
이와 같은 함수의 특성을 추상성이라 하며, 함수를 활용한 추상화를 적절하게 활용하는 것이
프로그래밍의 핵심에 포함된다.

### 도구로서의 함수
도구의 사용법을 알아야 하는 것처럼 함수의 경우는 
함수를 언제 어떻게 사용하는 가를 반드시 알아야 한다.
일반적으로 함수에 대해서는 이정도만 알고 있어도 충분하다.

### 새 함수 정의하기

앞서 언급한 함수들은 파이썬에서 이미 누군가에 의해 정의된 함수들이다.
이런 함수들을 파이썬 **내장함수**라 부른다.

하지만 일반 사용자가 필요에 따라 함수를 임의로 정의할 수 있다.
새 함수를 정의하는 양식은 다음과 같다.
```python
def 함수이름(인자1, 인자2, ...):
    코드
```

#### 예제

아래 시간을 분으로 변환해서 되돌려주는 함수를 아래와 같이 정의할 수 있다.

In [9]:
def hour_to_min(hour):
    minutes = hour * 60
    return minutes

`hour_to_min` 함수는 `hour`라는 매개변수 하나만을 사용하며 인자를 받으면 분 단위로 변경한 값을
담고 있는 `minutes` 변수의 값을 되돌려 준다.

함수가 되돌려 준 값은 다른 함수의 인자로 사용되거나 변수를 선언하는 값으로 활용될 수 있다. 
이와 같이 함수가 되돌려 주는 값을 필요에 따라 재활용할 수 있다.

예를 들어, 2시간이 몇 분인지 알고자 하면 `hour_to_min` 함수를 아래와 같이 사용하면 된다.

In [10]:
two_hour = hour_to_min(2)
print("두 시간은", two_hour, "분에 해당합니다.")

두 시간은 120 분에 해당합니다.


위 코드에서 `hour_to_min(2)`을 실행하여 되돌려 받은 값은 `two_hour` 변수에 할당하였고
그 변수를 `print` 함수의 둘째 인자로 활용하였다.

### 지역변수와 전역변수
함수를 선언할 때 사용되는 매개변수와 함수 본체에서 선언되는 변수는 함수가 실행되는 동안에만 의미를 갖는 변수들이며,
이런 변수들을 **지역변수**라 부른다. 
지역변수 가인 변수들은 **전역변수**라 부른다.

예를 들어, `hour_to_min` 함수를 정의할 때 사용된 `hour`와 본체에서 선언된 `minutes` 변수는 모두 지역함수이며,
`two_hour` 는 함수 밖에서 선언된 전역변수이다.

지역변수들은 함수 밖에서는 어떤 의미도 갖지 않는다. 예를 들어, 아래 코드를 실행하면 오류가 발생한다.

In [11]:
two_hour = hour_to_min(2)
print(minutes)

NameError: name 'minutes' is not defined

물론 아래의 경우도 오류가 발생한다.

In [12]:
two_hour = hour_to_min(2)
print(hour)

NameError: name 'hour' is not defined

위에서 오류가 발생하는 이유는 `hour_to_min` 함수가 인자 2와 함께 실행되어 종료가 되면 
실행도중에 선언되어 사용된 `hour`와 `minutes` 변수의 의미도 완전히 사라지기 때문이다.

아래 **함수호출**에 대한 설명에서 지역변수가 프로그램 실행 도중에 어떻게 의미를 갖거나 잃어버리는가를
**프레임** 개념을 이용하여 좀 더 자세히 설명될 것이다.

### 함수의 리턴값<a id='funReturn'></a>

`hour_to_min` 와 같은 함수가 되돌려주는 값을 일반적으로 **리턴(return)값**이라 부른다.
리턴 값은 함수 정의 본체에서 `return`이란 키워드 바로 뒤에 오는 값을 의미하며, 
함수가 특정 인자와 함께 실행되면 함수 본체의 코드에 맞추어 값을 조작한 후에 
최종적으로 되돌려주는 값을 가리킨다.

파이썬의 경우 리턴값이 앞서 설명한 대로 `return` 키워드와 함께 명시된 함수와 
그렇지 않은 함수가 존재한다. 

#### 리턴값이 명시된 함수
* `hour_to_min` 함수의 경우처럼 `return` 키워드가 명시적으로 사용된 함수.
* 함수를 실행하다가 `return`을 만나게 되면 리턴값을 되돌려 준 후 함수의 실행을 바로 멈춘다.

C 또는 Java 언어와는 달리 파이썬에서는 함수의 리턴값의 자료형을 명시하지 않는다.

#### 예제
`hour_to_min` 함수를 C언어로 구현한다면 다음과 같은 모양을 갖는다.
```c
int hour_to_min(int hour) {
    int minutes = hour * 60;
    return minutes;
}
```

#### 예제
`hour_to_min` 함수를 Java 언어로 구현한다면 다음과 같은 모양을 갖는다.
```java
public int hour_to_min(int hour) {
    int minutes = hour * 60;
    return minutes;
}
```

#### 리턴값이 명시되지 않은 함수
대표적으로 `print` 함수가 리턴값이 명시되지 않은 함수이다. 
리턴값이 명시되지 않은 함수의 리턴값은 항상 `None` 이다.
`None`은 사실상 아무런 의미를 갖지 않는 값이며 어떠한 재활용도 불가능하다.

예를 들어, 아래 코드는 `print` 함수가 리턴하는 값을 확인하고자 하지만 말했듯이 `None`이란 것만 확인할 수 있다.

In [13]:
x = print(1)
print(x)

1
None


**주의:** 위 코드를 실행한 결과의 첫째 줄에서 보이는 `1`은 `print` 함수의 리턴값이 아니라,
`print(1)`을 실행하여 모니터에 숫자 1을 출력하는 `print` **함수의 부수효과**에 불과하다.
함수의 부수효과에 대해서는 아래에서 좀 더 자세히 다룬다.

#### 예제
`hour_to_min` 함수를 아래와 같이 리턴값이 명시되지 않은 함수로 정의할 수도 있다.

In [14]:
def hour_to_min_noReturn(hour):
    minutes = hour * 60
    print(minutes)

위 함수를 이용하여 두 시간을 분단위로 확인하고자 하면 다음과 같다.

In [15]:
hour_to_min_noReturn(2)

120


위 코드를 보면 마치 120을 리턴하는 것처럼 보이지만 사실은 `print` 함수가 모니터에 출력한 결과를 
우리에게 보여주는 거에 불과하다.
실제로 아래와 같이 리턴값을 활용하려 하면 제대로 작동하지 않는다.

In [16]:
two_hour_noReturn = hour_to_min_noReturn(2)
print(two_hour_noReturn)

120
None


실제로 `hour_to_min_noReturn(2)`의 자료형은 `None`의 자료형인 `NoneType`이다.

In [17]:
type(hour_to_min_noReturn(2))

120


NoneType

#### 예제
`hour_to_min` 함수를 C언어로 구현한다면 다음과 같은 모양을 갖는다.
```c
void hour_to_min(int hour) {
    int minutes = hour * 60;
    printf(minutes);
}
```

#### 예제
`hour_to_min_noReturn` 함수를 Java 언어로 구현한다면 다음과 같은 모양을 갖는다.
```java
public void hour_to_min(int hour) {
    int minutes = hour * 60;
    System.out.println(minutes);
}
```

#### 함수의 리턴값은 단 하나

함수 본체 코드에 `return` 문이 여러 번 사용되었다 하더라도 함수는 하나의 값만 리턴한다.

예를 들어, 아래 루비에서 정의된 `login` 함수 본체에는 `return` 문이 두 번 사용되었다.

```ruby
def login(_id)
  members = ['egoing', 'k8805', 'leezche']
  for member in members do
      if member == _id
          return true
      end
  end
  return false
end
```

하지만 `login` 함수가 호출되어 실행되는 과정에 두 리턴문 문 중에 하나의 리턴문이 실행되면
바로 지정된 리턴값을 반환하면서 함수의 실행이 멈추게 되어 다른 리턴문은 실행되지 않게 된다.

## 함수 표현식

함수를 이용해서 값을 표현하는 방식 **함수 표현식**이라 부른다. 

예를 들어, 제곱근을 생성하는 함수인 `sqrt`의 경우 실수 하나를 인자로 받아서 그 실수의 제곱근을 생성하여 돌려준다.
`sqrt` 함수를 이용하여 실수 `r`의 제곱근을 아래와 같이 표현할 수 있다.
```python
sqrt(r)
```    
또한, 실수 `r`의 제곱근에다가 `2`를 더한 값은 다음과 같이 표현된다.
```python
add(sqrt(r),2)
```    

#### 주의사항
앞서 언급한 대로 `sqrt`를 이용한 함수 표현식이 `add` 함수의 첫째 인자로 활용되었다.
이런 표현식은 **함수의 합성**을 활용한 표현식이다.
이 경우 `add` 함수의 첫째 인자값으로 `sqrt(r)`이 실행되어 되돌려진 값, 즉, `r`의 제곱근이 사용된다.

## 함수호출

함수 표현식은 특정 값을 표현하기 위해 사용된다.
하지만 그 표현식이 어떤 값을 나타내는지를 알려면 사용된 함수들을 실행해야만 하며,
이렇게 함수 표현식을 실행하는 것을 **함수호출**이라 부른다.

예를 들어, `sqrt(8)`은 `8`의 제곱근을 나타내는 값이지만 실제로 `8`의 제곱근이 정확히 얼마인지는 말하지 않는다.
`8`의 제곱근이 얼마인지를 확인하기 위해서는 이제 `sqrt`에 `8`을 인자로 입력하여 호출해야 한다.

"`sqrt`에 `8`을 인자로 입력하여 호출한다" 란 말이 좀 복잡하게 들리지만 실제로는 매우 간단하다.
바로 아래 명령을 실행하기만 하면 된다.
```python
sqrt(8)
```

#### 주의사항
    
`sqrt`를 사용하려면 먼저 `math` 모듈을 임포트해야 한다. 
파이썬의 특정 모듈에서 특정 함수를 임포트 하는 방식은 다음과 같이 할 수 있다.
```python    
from math import sqrt
```

In [18]:
from math import sqrt

sqrt(8)

2.8284271247461903

즉, `sqrt(8)`의 정확한 값은 `2.8284271247461903`이다.
8의 제곱근이 무리수이기 때문에 근사값에 불과하다.
하지만 8의 제곱근의 정확한 값에 매우 근사한 값이다.

이와 같은 방식으로 `add(sqrt(r),2)`의 정확한 값을 계산할 수 있다.

In [19]:
add(sqrt(8), 2)

4.82842712474619

이와 같이 함수호출 표현식의 정확한 값을 계산하는 과정을 **함수호출**이라 부른다.
즉, 함수를 정말로 실행하여 특정 값을 얻어내는 과정이 함수호출이다.

### 함수호출 실행과정

함수호출 표현식을 실행하여 결과값을 확인하는 것을 함수호출이라 하였다.
여기서는 함수호출이 실행되는 과정을 예제들을 통해 좀 더 자세히 살펴본다. 

#### 예제

아래 함수호출 표현식의 결과값이 어떤 순서대로 정해지는가를 확인해보자.

**주의:** 
* `add` 함수와 마찬가지로 `mul` 함수는 일상적으로 곱셈(`*`) 기호로 표시되는 함수의 원래 이름이며,
    `operator` 모듈에 포함되어 있다.
* `0x`로 시작하는 숫자는 16진법을 의미한다.
* `0o`로 시작하는 숫자는 8진법을 의미한다.

    

In [20]:
from operator import mul

In [21]:
mul(add(2, mul(0x4, 0x6)), add(0x3, 0o5))

208

* `mul` 함수의 정의를 확인한다.
* `mul`의 첫 번째 인자인 `add(2, mul(0x4, 0x6))`를 확인하고 함수호출을 실행한다.
    + `add` 함수의 정의를 확인한다. 
    + `add`의 첫 번째 인자인 `2`를 확인한다. `2`는 더 이상 계산할 필요가 없다.
    + `add`의 두 번째 인자닌 `mul(0x4, 0x6)`를 확인하고 함수호출을 실행한다. 
        - `mul` 함수의 정의를 확인한다. 
        - `mul`의 첫 번째 인자인 `0x4`를 확인한다. `0x4`는 16진법으로 표시된 정수 4임을 확인한다.
        - `mul`의 두 번재 인자인 `0x6`를 확인한다. `0x6`는 16진법으로 표시된 정수 6임을 확인한다.
        - `mul(4,6)`을 실행하여 `24`를 리턴한다.
    + `add(2, 24)`를 실행하여 `26`을 리턴한다.
* `mul`의 두 번째 인자인 `add(0x3, 0o5)`를 확인한다. 
    + `add` 함수의 정의를 확인한다.
    + `add`의 첫 번째 인자인 `0x3`를 확인한다. `0x3`는 16진법으로 표시된 정수 3임을 확인한다.
    + `add`의 두 번째 인자인 `0o5`를 확인한다. `0o5`는 8진법으로 표시된 정수 5임을 확인한다.
    + `add(3, 5)`실행하여 `8`을 리턴한다.
* `mul(26, 8)`을 실행하여 `208`을 리턴한다.

위 설명에서 들여쓰기에 주의해야 한다.
더 들여서 쓴 문장이 모두 실행이 된 뒤에야 그 아랫부분에서 덜 들여서 쓴 문장이 실행된다. 
그림으로 나타내면 다음과 같다.

아래 그림에 네모상자로 둘러싸인 부분이 현재 실행중인 함수호출 또는 함수호출의 결과값을 나타낸다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/fun_call.png" style="width:500px">
</td>
</tr>
</table>
</p>

## 부수효과가 없는 함수와 있는 함수

프로그램에서 사용되는 함수를 구분하는 다양한 기준이 있다.
여기서는 부수기능의 존재여부에 따른 함수 분류 기준을 살펴본다.

### 부수효과가 없는 함수 (Pure functions)

함수가 호출되어 실행되어 실행결과를 돌려줄 때까지 실행결과를 생성하기 위한 일 이외에 다른 어떤 일도 하지 않는다면 그 함수를 부수효과가없는 함수라 부른다.

예를 들어 절대값을 계산하는 `abs` 함수와 덧셈을 행하는 `add` 함수 등이 순수한 함수이다. 
아래 그림에서 보듯이 인자를 입력받은 후에 각각 절대값과 덧셈을 실행하여 결과값을 돌려주는 일 이외에는 다른 일을 행하지 않는다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/fun_pure.png" style="width:250px">
</td>
</tr>
</table>
</p>

### 부수효과가 있는 함수 (Impure functions)

부수효과가 있는 함수는 부수효과가 없는 함수와는 달리 결과값을 돌려주는 것 이외에 부수적인 일을 한다.
대표적으로 파이썬 내장함수인 `print` 함수가 부수효과를 갖는 함수이다.

그런데 `print`가 부수효과를 갖는 함수라는 것을 확인하려면 아래 두 가지를 알아야 한다.

* 리턴하는 값이 무엇인가?
* 부수적으로 어떤 일을 하는가?

먼저, `print`가 리턴하는 값은 널 값(null value)이다. 
널 값은 말 그대로 어무런 내용도 없는 값을 의미한다.
즉, 값은 값인데 아무 데도 사용할 수 없는 값을 의미한다.
즉, 무늬만 값인 것이다. 
파이썬은 이러한 널값을 `None`으로 표시한다.

다음으로, `print` 함수가 부수적으로 하는 일은 예를 들어 터미널 창에 어떤 문자열을 출력하는 것이다.
`print("Hello Python")` 방식으로 `print` 함수를 호출하면 아래와 같이 `'Hello Python'` 이란 문자열을 출력한다.

In [22]:
print("Hello Python")

Hello Python


`print("Hello Python")`을 실행하였을 때 보여주는 `'Hello Python'`은 실행 결과값이 아니라 `print` 함수가 부수적으로 하는 일이라는 점을 절대적으로 명심해야 한다.

#### 예제: 부수효과를 갖는 함수의 호출과정

부수효과를 갖는 함수호출의 실행과정은 부수효과가 없는 함수의 호출과정과 기본적으로 동일하다.
하지만 중간중간에 부수적인 일도 함께 벌어진다.

먼저 아래 코드의 실행결과가 어떻게 도출되었는가를 잘 생각해보자.

In [23]:
print(print(1), print(2))

1
2
None None


아래의 그림은 위 코드를 호출하는 과정을 설명한다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/fun_print.png" style="width:350px">
</td>
</tr>
</table>
</p>

**설명:** 기본적으로 앞서 설명한 방식으로 진행되지만 3번 사용된 각각의 `print` 함수 호출과정에서 
특정값을 부수적으로 화면에 출력한다. 

* 가장 먼저 첫 번째 인자에 사용된 `print` 함수에 의해 `1`을 출력하고
* 다음으로 두 번째 인자에 사용된 `print` 함수에 의해 `2`를 출력하고
* 마지막으로 제일 먼저 호출된 `print` 함수에 의해 각 인자의 리턴값으로 이루어진 `None None`을 출력한다. 

## 함수호출 실행과정과 프레임 이해

함수를 호출하여 실행을 하면 컴퓨터 메모리의 스택(stack) 영역에 함수의 호출과 실행과정, 그리고 실행결과값과 관련된
정보의 변화가 일어난다. 
이런 정보의 변화를 **프레임(frame)** 이라 부르며, 프레임은 모든 함수와 변수의 생성, 호출, 삭제 과정을 기록한다.

또한 하나의 함수가 호출될 때마다 하나의 프레임이 임시로 생성되었다가 함수의 실행이 종료되면 해당 함수의
리턴값을 다른 프레임에 넘겨 준 후에 사라지는 과정이 반복된다. 이렇게 특정 함수호출과 관련된 프레임을 **지역 프레임**이라 
부른다.
반면에 프로그램이 실생되는 전 과정동안 살아 있는 프레임을 **전역 프레임**이라 부른다. 

아래 그림은 앞서 지역변수와 전역변수를 설명하면서 사용한 코드를 
[파이썬 튜터](http://pythontutor.com/)에서 실행한 결과를 보여준다.

#### 주의사항
* 전역변수와 함수들은 전연 프레임(global frame)에서 다룬다.
* 지역변수들은 함수가 호출되면 생성되어 지역변수들의 정보를 다루다가, 함수의 실행이 종료되면 
    모든 정보와 함께 사라진다. 
    예를 들어, 마지막 그림에서 `minutes` 값을 확인하고자 할 때 오류가 발생하는 이유가 
    `hour_to_min` 함수가 호출될 때 생성된 지역 프레임이 함수의 실행이 종료되면 사라지기 때문이라는 
    점을 눈으로 확인할 수 있다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var01.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var02.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var03.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var04.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var05.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var06.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var07.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var08.png" style="width:90%">
</td>
</tr>
</table>
</p>

## 예제: 재귀함수

올해의 게임(Game Of The Year, GOTY)을 담고 있는 리스트가 아래와 같이 있다.

In [ ]:
GOTY = ["Zelda", 2017, "Uncharted", 2016, 
           ["Farcry5", 
            ["A Way Out", "God of War", "Detroit"]]]

위 리스트를 아래와 같이 출력하는 코드는 2중 `for`문을 이용할 수 있다.
```
Zelda
2017
Uncharted
2016
Farcry5
['A Way Out', 'God of War', 'Detroit']
```

In [ ]:
for each_item in GOTY:
    if isinstance(each_item, list):
        for nested_item in each_item:
            print(nested_item)
    else:
        print(each_item)

이제 `GOTY`를 아래 모양으로 출력하는 코드를 구현하고자 한다.
즉, 중첩으로 사용된 리스트를 모두 해체하는 코드가 필요하다.
```
Zelda
2017
Uncharted
2016
Farcry5
A Way Out
God of War
Detroit
```

가장 먼저 떠오르는 것은 아마도 `for`문을 3중으로 겹쳐서 사용하면 된다는 것이다.
즉, 아래와 같이 할 수 있다.

In [ ]:
for each_item in GOTY:
    if isinstance(each_item, list):
        for nested_item in each_item:
            if isinstance(nested_item, list):
                for double_nested_item in nested_item:
                    print(double_nested_item)
            else:
                print(nested_item)
    else:
        print(each_item)

그런데 이렇게 하면 한 가지 문제가 있음을 알아챘을 것이다. 
`GOTY` 리스트의 중첩 정도와 원하는 결과에 따라 사용해야 하는 `for`문의 중첩 정도가 상응해서 복잡해지기 때문이다.
현재 `moviesw` 리스트는 3중 리스트 모양이다. 따라서 가장 안쪽에 위치한 리스트의 내용을 확인하려면 
`for`문을 3중으로 이용해야 한다.
그렇다면 4중, 5중, 6중 등등의 리스트를 다룬다면 4중, 5중, 6중 등등의 `for`문을 사용해야 한다는 말인데, 
이건 매우 심각하다. 

이와같이 대상으로 삼는 리스트의 중첩 정도에 따라 다른 식으로 구현된 프로그램은 최악 중의 최악이다. 
도대체 몇 중의 리스트를 다룰지 매번 보고 결정해야 한다는 말인데 매우 비효율적이며 
어떠한 실용성도 없는 프로그램 작성법이다. 

그렇다면 리스트의 중첩 정도의 정보를 굳이 사용하지 않으면서 원하는 결과를 얻어낼 수 있는 방법이 필요하다. 
여기서 재귀함수를 이용한 해결책을 소개한다. 

먼저 다음에 정의되는 `print_items` 함수를 살펴보자.

In [ ]:
def print_items(a_list):
    for each_item in a_list:
        if isinstance(each_item, list):
            print_items(each_item)
        else:
            print(each_item)

`print_items` 함수는 좀 이상하다. 
정의가 끝나지 않았는데 자신을 자신 본체에서 사용한다. 
이런 함수를 __재귀함수__(recursive function)라 부른다. 

4번 줄에서 사용된 `print_items` 함수호출이 이루어지면 다시 2번 줄로 돌아간다.
즉, 현재 확인하는 항목이 또다른 리스트인지를 묻는 질문을 리스트가 아닐 때까지 반복한다.
따라서 `print_items` 함수를 이용하면 앞선 문제를 바로 해결한다.
사실 임의로 중첩된 리스트를 인자로 받아도 중첩을 모두 풀어버린다.

In [ ]:
print_items(GOTY)

## 연습문제

1. `print_items` 수정하여 `GOTY`를 아래와 같이 출력하도록 하는 `print_items2` 함수를 구현하라.

        Zelda
        2017
        Uncharted
        2016
            Farcry5
                A Way Out
                God of War
                Detroit

    ##### 힌트
    * `print_items` 함수의 인자를 두 개로 수정한다. 
        하나는 리스트의 인자를 다루며, 다른 하나는 들어쓰기 정도를 다루는 
        인자를 하나 받아서 앞 문제 코드의 for문을 재귀적으로 처리한다. 
        즉, 아래와 같은 모양을 갖는다.

            def print_lol2(a_list, level):
                함수본체

        위에서 level은 탭을 사용하는 횟수를 나타내도록 한다. 
        그러면 `print_items2(GOTY, 0)`을 실행하면 원하는 결과가 나올 수 있다.
    * 탭 출력은 `print('\t')`를 이용하면 된다.
    <br><br>
1. 위 과제에서 구현한 `print_items2` 함수를 아래와 같이 수정하라.
    * 인자수를 세 개로 늘린다.

            def print_items3(a_list, level, indent=False):
                함수본체

    * `indent` 예약어인자 값이 `True`이면 연습4에서 처럼 들여쓰기를 하고, 
        `False`이면 들여쓰기를 하지 않는다.

## 연습문제
1. 다음 조건을 만족시키는 `right_justify`라는 함수를 정의하라.
    * 인자는 하나만 받으며, 매개변수는 `s`라 부른다.
    * 문자열 하나를 인자로 받아 실행하면 해당 문자열의 끝이 70번째 칸에 위치하도록 
        입력받은 문자열 앞에 충분한 공백이 위치하도록 출력(print)한다. 
```python
>>> right_justify('monty')
                                                                monty
```                                                                 
    힌트: 문자열 결합 및 반복, 그리고 문자열의 길이를 되돌려주는 내장함수 `len` 활용 가능.
<br><br>
1. 파이썬은 정의된 함수도 하나의 값으로 취급한다. 
    따라서 함수를 다른 함수의 인자로 사용하거나 변수에 할당되는 값으로 사용될 수 있다. 
    예를 들어, 아래에 정의된 `do_twice` 함수는 함수 `f`를 인자로 입력받으면
    그 함수를 두 번 호출하여 실행하도록 하는 함수이다. 
```python
def do_twice(f):
    f()
    f()
```
    그리고 `print_spam` 이라는 함수를 두 번 호출하도록 `do_twice`를 활용하고자 하면 
    아래와 같이 프로그램을 작성하면 된다.
    ```python
    def print_spam():
        print('spam')

    do_twice(print_spam)
    ```
    1. 이 프로그램을 직접 입력해서 실행해 보라.
    1. 아래 조건을 만족하도록 `do_twice` 함수를 수정하라. 
        * 두 개의 인자를 사용한다.
        * 첫째 매개변수는 하나의 인자를 받는 함수를 인자로 입력받는다.
        * 둘째 매개변수는 첫째 인자에 사용된 함수의 인자로 사용될 수 있는 값을 인자로 입력받는다.
        * 첫째 인자로 사용된 함수를 둘째 인자로 사용된 값을 이용하여 두 번 연속 호출한다. 
    1. 아래 조건을 만족하도록 `do_four` 함수를 수정하라. 
        * 두 개의 인자를 사용한다.
        * 첫째 매개변수는 하나의 인자를 받는 함수를 인자로 입력받는다.
        * 둘째 매개변수는 첫째 인자에 사용된 함수의 인자로 사용될 수 있는 값을 인자로 입력받는다.
        * 첫째 인자로 사용된 함수를 둘째 인자로 사용된 값을 이용하여 네 번 연속 호출한다. 
        * 앞서 정의된 `do_twice` 함수를 반드시 활용한다.
<br><br>
답: http://greenteapress.com/thinkpython2/code/do_four.py
<br><br>
1. 이번 문제는 지금까지 배운 내용을로 풀 수 있다. 
    1. 다음과 같은 격자를 그리는 함수를 작성하라.
        ```
        + - - - - + - - - - +
        |         |         |
        |         |         |
        |         |         |
        |         |         |
        + - - - - + - - - - +
        |         |         |
        |         |         |
        |         |         |
        |         |         |
        + - - - - + - - - - +
        ```
        힌트: 한 줄에 하나 이상의 값을 출력하려면 `print` 함수에 여러 인자를 사용하면 된다.
        ```python
        print('+', '-')
        ```
        `print` 함수는 기본적으로 줄바꿈을 실행한다.
        줄바꿈을 하지 않기 위해서는 에를 들어 아래와 같이 실행하면 된다.
        ```
        print('+', end=' ')
        print('-')
        ```
        위 프로그램을 실행하면 `+ -` 형식으로 출력된다. 
        <br><br>
    1. 4개의 행과 4개의 열을 갖는 비슷한 격자를 그리는 함수를 작성하라.

    답: http://greenteapress.com/thinkpython2/code/grid.py
    <br><br>
    이 연습문제는 아래 책의 Oualline에 포함된 연습문제를 응용하였다. 
    
            Practical C Programming, Third Edition, O’Reilly Media, 1997.